tutorials/python/python-type-driven-development/tutorial.md

```diff
@@ -0,0 +1,207 @@
+
+
+print(make_hamburger("ground beef", 2))
+# ['bread', 'ground beef', 'ground beef', 'bread']
+```
+
+Let’s see how it goes with types:
+
+```python
+from typing import List
+
+
+def make_hamburger(meat: str, number_of_meats: int) -> List[str]:
+   return ["bread"] + [meat] * number_of_meats + ["bread"]
+```
+
+What happened here? We defined the function’s arguments types: string for `meat`, integer for `number_of_meats` and the return type which is a list of string values. Did you see the import at the beginning? Complex types must be imported from the `typing` library. We will cover some of them later.
+
+## Type Alias
+
+You can also define your custom type names for known structures, very useful to improve the readability. Let’s say if we want to consider a hamburger is a list of strings, we can define a type `Hamburger` in the following way:
+
+```python
+Hamburger = List[str]
+
+def make_hamburger(meat: str, number_of_meats: int) -> Hamburger:
+   return ["bread"] + [meat] * number_of_meats + ["bread"]
+```
+
+## Callables
+
+Are you doing functional programming? No problem, the typing library inclues a type for callables which accepts a two-dimensional list like `[[argument1_type, … argumentN_type], return_type]`.
+
+```python
+from typing import Callable, Optional
+
+
+def sum_and_process(a: int, b: int, callback: Callable[[int, Optional[str]], bool]) -> bool:
+   total = a + b
+   return callback(total, None)
+
+
+def is_positive(val: int, message: Optional[str]) -> bool:
+   if message:
+       print(message)
+   return val > 0
+
+
+output = sum_and_process(5, 2, is_positive)
+print(output)
+# True
+```
+
+First of all, we are creating a function that takes two integer values, sums up them and returns the output of an incoming callback. This callback `is_positive` is defined as a function that takes one mandatory integer and one optional string. If the message exists it will print it, and always returns if the incoming value is positive. Of course, this function is not pure because of the print side-effect, but it’s just an example :)
+
+## Generics and Union Types
+
+One of the most powerful features in a type system is the generics and union types. It’s also possible to use these in Python annotations. Let’s take a look at the following example:
+
+```python
+from typing import TypeVar, List
+
+
+T = TypeVar("T", int, List[str])
+
+def generic_add(x: T, y: T) -> T:
+   return x + y
+
+
+x1: int = 5
+y1: int = 2
+print(generic_add(x1, y1))
+# 7
+
+x2: List[str] = ["Hello"]
+y2: List[str] = ["World"]
+print(generic_add(x2, y2))
+# ['Hello', 'World']
+
+x3: str = "foo"
+y3: str = "bar"
+print(generic_add(x3, y3))
+# error: Value of type variable "T" of "generic_add" cannot be "str"
+```
+
+Above, we defined a type variable `T` that can be integer or List of strings and a `generic_add` function that perform the “addition” operation for the incoming arguments which are limited by the `TypeVar` declaration.
```

Comment:
I think that this paragraph could say something more about the example or at least about he error at the end. Right now it looks like it only mentions the beginning of the example.